Firestore/core/src/firebase/firestore/remote/grpc_connection.mm

```diff
@@ -0,0 +1,123 @@
+#include "Firestore/core/src/firebase/firestore/model/database_id.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+#include "Firestore/core/src/firebase/firestore/util/string_format.h"
+#include "absl/memory/memory.h"
+#include "grpcpp/create_channel.h"
+
+#import "Firestore/Source/API/FIRFirestoreVersion.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+using core::DatabaseInfo;
+using model::DatabaseId;
+using util::StringFormat;
+
+namespace {
+
+const char *const kXGoogAPIClientHeader = "x-goog-api-client";
+const char *const kGoogleCloudResourcePrefix = "google-cloud-resource-prefix";
+
+std::string MakeString(absl::string_view view) {
+  return view.data() ? std::string{view.data(), view.size()} : std::string{};
+}
+
+}  // namespace
+
+GrpcConnection::GrpcConnection(const DatabaseInfo &database_info,
+                               util::AsyncQueue *worker_queue,
+                               grpc::CompletionQueue *grpc_queue)
+    : database_info_{&database_info},
+      worker_queue_{worker_queue},
+      grpc_queue_{grpc_queue},
+      grpc_channel_{CreateChannel()},
+      grpc_stub_{grpc_channel_} {
+}
+
+std::unique_ptr<grpc::ClientContext> GrpcConnection::CreateContext(
+    absl::string_view token) const {
+  auto context = absl::make_unique<grpc::ClientContext>();
+  if (token.data()) {
+    context->set_credentials(grpc::AccessTokenCredentials(MakeString(token)));
+  }
+
+  // TODO(dimond): This should ideally also include the grpc version, however,
+  // gRPC defines the version as a macro, so it would be hardcoded based on
+  // version we have at compile time of the Firestore library, rather than the
+  // version available at runtime/at compile time by the user of the library.
+  //
+  // TODO(varconst): this should be configurable (e.g., "gl-cpp" or similar for
+  // C++ SDK, etc.).
+  context->AddMetadata(
+      kXGoogAPIClientHeader,
+      StringFormat("gl-objc/ fire/%s grpc/",
+                   reinterpret_cast<const char *>(FIRFirestoreVersionString)));
+
+  // This header is used to improve routing and project isolation by the
+  // backend.
+  const DatabaseId &db_id = database_info_->database_id();
+  context->AddMetadata(kGoogleCloudResourcePrefix,
+                       StringFormat("projects/%s/databases/%s",
+                                    db_id.project_id(), db_id.database_id()));
+  return context;
+}
+
+void GrpcConnection::EnsureActiveStub() {
+  // TODO(varconst): find out in which cases a gRPC channel might shut down.
+  // This might be overkill.
+  if (!grpc_channel_ || grpc_channel_->GetState(/*try_to_connect=*/false) ==
+                            GRPC_CHANNEL_SHUTDOWN) {
+    LOG_DEBUG("Creating Firestore stub.");
+    grpc_channel_ = CreateChannel();
+    grpc_stub_ = grpc::GenericStub{grpc_channel_};
+  }
+}
+
+std::shared_ptr<grpc::Channel> GrpcConnection::CreateChannel() const {
```

Comment:
Note: `grpc::GenericStub` doesn't have a default constructor, and I'm not sure if it is guaranteed to work with a `null` channel; therefore, the stub has to be initialized in the constructor, which necessitates this helper for channel. Alternatively, stub could be stored in a `unique_ptr`, in which case this function could be merged with `EnsureActiveStub`.

Reply:
We likely should defer channel and stub creation if possible. We need more data to be sure of this but on Android at least we saw non-trivial cost in creating stubs that got in the way of serving the first query results from cache.